Add portable Dicom.Json as a separate nuget package. Connected to #182

rickardraysearch opened the pull request:
There are some issues with this PR, that I would like to discuss before merging. 

Firstly, the BulkUriByteBuffer implementation automatically downloads the referenced bulk data using a web request if the buffer is accessed, which has some security implications. Does anyone have a better idea on how to do it? One option could be to throw exceptions when accessing the buffer until an explicit download method has been invoked, similar to what is done in BulkUriByteBuffer.Size. Would that be better?

Secondly, the conversion from json to Dicom of DS attributes does not preserve the entire string, since it seems Json.NET doesn't allow accessing decimal values as strings, but always parses them into doubles or decimals. Do you see that as a problem?

rickardraysearch commented:
The failing test is Dicom.Network.DicomClientTest.SendAsync_MultipleTimes_AllRecognized(expected: 20) which seems to have some reliability issues - it sometimes fails locally as well.

anders9ustafsson commented:
@rickardraysearch Many thanks for this quick and solid PR. I will go through it in more detail as soon as possible. I would like to schedule this functionality for 2.1, I hope you are OK with that.

Regarding the failing test, yes, this part is not yet fully stable, but it performs considerably better than 1.1. I would say that it does not represent a common use case, but ideally it should of course never fail. The networking part could and should be further refactored, no doubt about that.

IanYates commented on DICOM/IO/Buffer/BulkUriByteBuffer.cs:
```diff
@@ -0,0 +1,99 @@
+      {
+        if (_buffer == null)
+          throw new InvalidOperationException(
+            "BulkUriByteBuffer cannot provide Data until either GetData() has been called.");
```
"either"...   (maybe GetByteRange() is meant to be here?)

rickardraysearch replied:
Yes, sorry about that. I was distracted by that GetByteRange fetches the data too, whereas only accessing the Data property does not. Sometimes SYNTAX ERRORs for English would be useful...

IanYates replied:
No worries - it's only because I read this spot a few times that I noticed it :)

IanYates commented on DICOM/IO/Buffer/BulkUriByteBuffer.cs:
```diff
@@ -0,0 +1,99 @@
+    {
+      if (_buffer != null) return _buffer;
+
+      using (var response = WebRequest.Create(BulkDataUri).GetResponseAsync().Result)
```
In desktop-land there's a non-async .Get().  I guess we're aiming for portability so .Result on the async method is probably the way to go.  But if we're going portable (and I'm not in that space so could be wrong), is HttpClient easier to use or is at least better supported?  I don't mind personally...

rickardraysearch replied:
Since the GetData method isn't defined in IByteBuffer, this could be changed into a GetDataAsync instead. I'll include this in the updated PR.

IanYates replied:
OK :+1:

IanYates commented:
@rickardraysearch - impressive work!  Very minor nit-picks, and possible non-issues, in my comments.

I took a brief look through the code to see what we could do about allowing the async nature of Data, ByteRange and Size on the ByteBuffer to be exposed but it'd be a HUGE breaking change so if we need .Wait() on an async call to maintain portability then that's cool with me.

Good thinking about the potential security / DOS implications about doing the download.  I'm not very familiar with the dicom/json specification being implemented here (although the tests helped make a lot of it quite clear) so this question may seem silly - apologies if so.  Is there a need to allow interception of the HTTP download process anyway so that things like HTTP headers, or other authentication concerns, can be addressed by the client-side library?  If that's the case then that'd be the ideal place to also allow for screening of the download.

rickardraysearch commented:
Thanks for the kind words, @IanYates! Your suggestion on intercepting HTTP download mechanism needs to be addressed. Perhaps the straight-forward way is to completely remove downloading from this class, and only keep a method for setting the buffer, like this:
```
public void SetBuffer(byte[] buffer) { _buffer = buffer; }
```
It would then be reasonable to throw exceptions on both accessing Data and GetByteRange.

Perhaps there should be a BulkUriByteBufferFactory on JsonDicomConverter, for allowing subclasses with custom downloading behaviour? Then users could implement GetByteRange with HTTP Ranges headers, custom caching, etc.

rickardraysearch commented:
I have refactored the BulkData things to allow customizing the ByteBuffers created by subclassing JsonDicomConverter and overriding CreateBulkDataUriByteBuffer, similar to how it's done in DicomServer.CreateScp. 

I also removed all WebRequest code from BulkDataUriByteBuffer, delegating that to application implementations. There is still an example in JsonDicomConverterTest.DownloadBulkData that is used in JsonDicomConverterTest.BulkDataRead.

Is this better than the old approach?

anders9ustafsson commented:
Great work, @rickardraysearch ! I think I'll leave it to @IanYates to have an opinion on the web related parts, but I have a few minor comments on other stuff:

* In the *DICOM.Json* library, please include the *SharedAssemblyInfo.cs* for common assembly attributes.
* In the new *.nuspec* file, set `target` for the .dll and .pdb to *lib\portable-net45+netcore45+wpa81* (instead of *lib\net45*) to ensure that the package can be accessed by all the supported platform targets.

IanYates commented:
@rickardraysearch - seems great to me!

anders9ustafsson commented:
@rickardraysearch Terrific! One final thing: could you please add simple API documentation on all `public` methods and properties in `BulkDataUriByteBuffer` and `JsonDicomConverter`?

API documentation is already lagging behind, but I want us to keep our heads over the water and at the very least include API documentation when the public API is extended.

rickardraysearch commented:
@anders9ustafsson I completely agree about the documentation. I'll try to get it fixed tonight.
